File: c++/c25_3_sprague_grundy_theorem.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	template<class T, class U>
	map<T, ll> grundy(const vector<T>& states, U func) {
		map<T, ll> cache;
		auto rec = [&](this auto self, const T& s) -> ll {
			if (cache.contains(s))
				return cache[s];
			auto n = func(s) | views::transform([&](const T& i) { return self(i); }) | ranges::to<unordered_set>();
			for (ll i = 0; ; ++i) {
				if (!n.contains(i)) {
					cache[s] = i;
					return i;
				}
			}
			};
		for (const auto& s : states)
			rec(s);
		return cache;
	}

	vector<vector<ll>> grundy_maze(const vector<string>& m) {
		auto maze = [&](const pair<ll, ll>& s) {
			auto [y, x] = s;
			vector<pair<ll, ll>> ret;
			for (ll y1 = y - 1; y1 >= 0; --y1) {
				if (m[y1][x] == '#')
					break;
				ret.emplace_back(y1, x);
			}
			for (ll x1 = x - 1; x1 >= 0; --x1) {
				if (m[y][x1] == '#')
					break;
				ret.emplace_back(y, x1);
			}
			return ret;
			};
		ll h = ssize(m), w = ssize(m[0]);
		vector<vector<ll>> r(h, vector<ll>(w, -1));
		vector<pair<ll, ll>> states;
		for (ll i = 0; i < h; ++i) {
			for (ll j = 0; j < w; ++j) {
				if (m[i][j] != '#') {
					states.emplace_back(i, j);
				}
			}
		}
		for (const auto& [p, v] : grundy(states, maze)) {
			auto [y, x] = p;
			r[y][x] = v;
		}
		return r;
	}
}

TEST(C253SpragueGrundyTheorem, grundy) {
	EXPECT_EQ(grundy<ll>({ 0,1,2,3,4,5 }, [](ll s) { return vector<vector<ll>>{{},{0},{1},{0,1},{1,3},{1,4}}[s]; }), (map<ll, ll>{{0,0},{1,1},{2,0},{3,2},{4,0},{5,2}}));
}

TEST(C253SpragueGrundyTheorem, grundyMaze) {
	EXPECT_EQ(grundy_maze({"..#..","#...#","..#..","#....","....."}), (vector<vector<ll>>{{0,1,-1,0,1},{-1,0,1,2,-1},{0,2,-1,1,0},{-1,3,0,4,1},{0,4,1,3,2}}));
}

File: python/c25_3_sprague_grundy_theorem.py
import unittest
from itertools import count
import sys

sys.setrecursionlimit(max(sys.getrecursionlimit(),1<<20))

def grundy(states,func):
    cache={}
    def rec(s):
        if s in cache:
            return cache[s]
        n=set(rec(i) for i in func(s))
        for i in count():
            if i not in n:
                cache[s]=i
                return i
    for s in states:
        rec(s)
    return cache

def grundy_maze(m):
    def maze(s):
        y,x=s
        for y1 in range(y-1,-1,-1):
            if m[y1][x]=='#':
                break
            yield y1,x
        for x1 in range(x-1,-1,-1):
            if m[y][x1]=='#':
                break
            yield y,x1
    h,w=len(m),len(m[0])
    r=[[None]*w for _ in range(h)]
    for (y,x),v in grundy(((i,j) for i in range(h) for j in range(w) if m[i][j]!='#'),maze).items():
        r[y][x]=v
    return r

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual(grundy(range(6),lambda s:[[],[0],[1],[0,1],[1,3],[1,4]][s]),{0:0,1:1,2:0,3:2,4:0,5:2})
        self.assertEqual(grundy_maze(['..#..','#...#','..#..','#....','.....']),[[0,1,None,0,1],[None,0,1,2,None],[0,2,None,1,0],[None,3,0,4,1],[0,4,1,3,2]])

if __name__=='__main__':
    unittest.main()
